### PR TITLE
Let (whitelisted) ghosts join in as Working Joes

### PR DESCRIPTION
**Pull request by victorjosephespinoza**
# About the pull request
https://forum.cm-ss13.com/t/make-it-so-ghosted-players-can-spawn-in-as-working-joes/588
Adds a ghost verb so ghosts can now join the game as working joes.

# Explain why it's good for the game
Like it was said in the forum post, Joes are pretty much harmless, their impact on the round is pretty much minimal, so adding this wouldn't imply any balance change whatsoever.

The intent of adding something like this to the server is mainly trying to get more attention to the role. Which is barely played at the moment, however, if dead players could now respawn as them, I believe it might get more activity.

# Testing Photographs and Procedure

Unsure of what kind of debugging I could provide in here, request any in the comments and I'll do it.

There's a high chance that something is not treated as it should. This is probably the first PR where I had to look and understand what I was writing, despite the fact that a lot of the code is taken from the pred initialization. So yeah, if you would have placed something somewhere else, or differently, go ahead and tell me.

# Changelog

:cl:H20Begod
add: Added a new verb for ghosts to join in as working joes (ONLY WHITELISTED GHOSTS).
code: Edited max spawns of the working joes to 3.
/:cl:

### CODE REVIEW

**Comment by cm13live** [comment]
This pull request has been mentioned on **CM-SS13**. Please provide feedback here:

https://forum.cm-ss13.com/t/make-it-so-ghosted-players-can-spawn-in-as-working-joes/588/6

**Review by morrowwolf — CHANGES_REQUESTED** [review]
Drop total positions for working joe to 3. I don't want to see a horde of working joes every hijack.

File: code/game/gamemodes/cm_initialize.dm [hunk 1]
```diff
@@ -830,3 +831,80 @@ Additional game mode variables.
...
+	msg_admin_niche("Ghost ([new_joe.key]) has joined as Working Joe, [new_joe.real_name].")
+
+	if(joe_candidate) joe_candidate.moveToNullspace() //Nullspace it for garbage collection later.
+
+/datum/game_mode/proc/check_joe_late_join(mob/joe_candidate, show_warning = 1)
+
+	if(!joe_candidate.client)
+		return
+
+	var/datum/job/J = RoleAuthority.roles_by_name[JOB_WORKING_JOE]
```
Comment (harryob) [inline code/game/gamemodes/cm_initialize.dm:858]: J -> joe_job

File: code/game/gamemodes/cm_initialize.dm [hunk 2]
```diff
@@ -830,3 +831,80 @@ Additional game mode variables.
...
+/datum/game_mode/proc/initialize_joe(mob/living/carbon/human/joe)
+	joes += joe.ckey
+
+/datum/game_mode/proc/attempt_to_join_as_joe(mob/joe_candidate)
+	var/mob/living/carbon/human/new_joe = transform_joe(joe_candidate) //Initialized and ready.
+	if(!new_joe) return
+
+	msg_admin_niche("Ghost ([new_joe.key]) has joined as Working Joe, [new_joe.real_name].")
+
+	if(joe_candidate) joe_candidate.moveToNullspace() //Nullspace it for garbage collection later.
```
Comment (harryob) [inline code/game/gamemodes/cm_initialize.dm:851]:
```suggestion
	if(joe_candidate) 
		joe_candidate.moveToNullspace() //Nullspace it for garbage collection later.
```

File: code/game/gamemodes/cm_initialize.dm [hunk 3]
```diff
@@ -830,3 +831,80 @@ Additional game mode variables.
...
+				//JOE INITIATLIZE\\
+
+//===================================================\\
+
+/datum/game_mode/proc/initialize_joe(mob/living/carbon/human/joe)
+	joes += joe.ckey
+
+/datum/game_mode/proc/attempt_to_join_as_joe(mob/joe_candidate)
+	var/mob/living/carbon/human/new_joe = transform_joe(joe_candidate) //Initialized and ready.
+	if(!new_joe) return
```
[the remark this answers is not in the document]
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: Changed in 03f51eb

File: code/game/gamemodes/cm_initialize.dm (hunk 2 above)
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: Changed in 03f51eb

File: code/game/gamemodes/cm_initialize.dm (hunk 1 above)
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: Changed in 03f51eb

File: code/game/gamemodes/cm_initialize.dm [hunk 4]
```diff
@@ -47,6 +47,7 @@ Additional game mode variables.
 	var/datum/mind/hellhounds[] = list() //Hellhound spawning is not supported at round start.
 	var/list/dead_queens // A list of messages listing the dead queens
 	var/predators = list()
+	var/joes	  = list()
```
[the remark this answers is not in the document]
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: Changed in 03f51eb

Coming from python this shocked me 😅

File: code/game/gamemodes/cm_initialize.dm [hunk 5]
```diff
@@ -830,3 +831,80 @@ Additional game mode variables.
 
 /datum/game_mode/proc/get_escape_menu()
 	return "On the [SSmapping.configs[SHIP_MAP].map_name], orbiting..."
+
+
+//===================================================\\
+
+				//JOE INITIATLIZE\\
```
[the remark this answers is not in the document]
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: Changed in 03f51eb

**Comment by victorjosephespinoza** [comment]
> Drop total positions for working joe to 3. I don't want to see a horde of working joes every hijack.

302807c8f59f8db8cda9727c6c5348320a685919

File: code/game/jobs/job/civilians/support/working_joe.dm [hunk 6]
```diff
@@ -1,7 +1,7 @@
 /datum/job/civilian/working_joe
 	title = JOB_WORKING_JOE
-	total_positions = 8
-	spawn_positions = 8
+	total_positions = 3
+	spawn_positions = 3
```
Comment (realforest2001) [inline code/game/jobs/job/civilians/support/working_joe.dm:4]:
```suggestion
	total_positions = 8
	spawn_positions = 8
```
Not sure why you're decreasing max slots. Currently it's never hit that amount, and you've not documented it in Changelog.

**Review by realforest2001 — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/cm_initialize.dm [hunk 7]
```diff
@@ -46,7 +46,8 @@ Additional game mode variables.
 	var/datum/mind/CO_survivor = null
 	var/datum/mind/hellhounds[] = list() //Hellhound spawning is not supported at round start.
 	var/list/dead_queens // A list of messages listing the dead queens
-	var/predators = list()
+	var/predators	= list()
```
[the remark this answers is not in the document]
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: ccce316056e2a0fbb201717f9c337260d93bf33e

File: code/game/jobs/job/civilians/support/working_joe.dm (hunk 6 above)
Comment (morrowwolf) [reply code/game/jobs/job/civilians/support/working_joe.dm:4]: I asked for the total to be lowered. Just add it to the changelog.

Comment (victorjosephespinoza) [reply code/game/jobs/job/civilians/support/working_joe.dm:4]: I've added it, forgot about it previously. Should be good now

File: code/game/gamemodes/cm_initialize.dm [hunk 8]
```diff
@@ -830,3 +831,81 @@ Additional game mode variables.
...
+
+	var/datum/job/joe_job = RoleAuthority.roles_by_name[JOB_WORKING_JOE]
+
+	if(!joe_job)
+		if(show_warning) to_chat(joe_candidate, SPAN_WARNING("Something went wrong!"))
+		return
+
+	if(!(RoleAuthority.roles_whitelist[joe_candidate.ckey] & WHITELIST_JOE))
+		if(show_warning) to_chat(joe_candidate, SPAN_WARNING("You are not whitelisted! You may apply on the forums to be whitelisted as a synth."))
+		return
```
Comment (harryob) [inline code/game/gamemodes/cm_initialize.dm:868]:
```suggestion
		if(show_warning) 
			to_chat(joe_candidate, SPAN_WARNING("You are not whitelisted! You may apply on the forums to be whitelisted as a synth."))
		return
```

File: code/game/gamemodes/cm_initialize.dm [hunk 9]
```diff
@@ -830,3 +831,81 @@ Additional game mode variables.
...
+			return
+
+	if(!enter_allowed)
+		if(show_warning) to_chat(joe_candidate, SPAN_WARNING("There is an administrative lock from entering the game."))
+		return
+
+	if(show_warning && tgui_alert(joe_candidate, "Confirm joining as a Working Joe.", "Confirmation", list("Yes", "No"), 10 SECONDS) != "Yes")
+		return
+
+	return 1
```
Comment (harryob) [inline code/game/gamemodes/cm_initialize.dm:889]:
```suggestion
	return TRUE
```

**Review by harryob — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/cm_initialize.dm (hunk 9 above)
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm:890]: bd073546ade9dcea2d13ea8a468026f565b9acff

**Comment by nauticall** [comment]
might want to clarify in the PR & changelog that WJs are restricted to synth whitelist holders

**Comment by victorjosephespinoza** [comment]
> might want to clarify in the PR & changelog that WJs are restricted to synth whitelist holders

woops, yeah, added it to the CL the other day but didn't do it on the title, doing that now

File: code/game/gamemodes/cm_initialize.dm [hunk 10]
```diff
@@ -830,3 +831,82 @@ Additional game mode variables.
...
+		return
+
+	var/datum/job/joe_job = RoleAuthority.roles_by_name[JOB_WORKING_JOE]
+
+	if(!joe_job)
+		if(show_warning) to_chat(joe_candidate, SPAN_WARNING("Something went wrong!"))
+		return
+
+	if(!(RoleAuthority.roles_whitelist[joe_candidate.ckey] & WHITELIST_JOE))
+		if(show_warning) to_chat(joe_candidate, SPAN_WARNING("You are not whitelisted! You may apply on the forums to be whitelisted as a synth."))
```
Comment (morrowwolf) [inline code/game/gamemodes/cm_initialize.dm:867]: This seems to have reverted to the same line again. Please move the to_chat() to the next line.

**Review by morrowwolf — CHANGES_REQUESTED** [review]

File: code/game/gamemodes/cm_initialize.dm (hunk 10 above)
Comment (victorjosephespinoza) [reply code/game/gamemodes/cm_initialize.dm]: 53ebddda58658b0f8c7e0bdfa913b7595d15426e

**Review by morrowwolf — APPROVED** [review]

**Pull request merged**